compiler/compiler.go

```diff
@@ -0,0 +1,213 @@
+	switch {
+	case e.Field != nil:
+		b.buildField(e.Field)
+	case e.Message != nil:
+		m := newMessage(e.Message, b.proto3, b.fullName)
+		b.messageDesc.NestedType = append(b.messageDesc.NestedType, m)
+	case e.Enum != nil:
+	case e.Option != nil:
+	case e.Oneof != nil:
+	case e.Extend != nil:
+	case e.Reserved != nil:
+	case e.Extensions != nil:
+	default:
+		panic("cannot interpret MessageEntry")
+	}
+}
+
+func (b *messageBuilder) buildField(pField *parser.Field) {
+	fieldDesc := b.createField(pField)
+	b.messageDesc.Field = append(b.messageDesc.Field, fieldDesc)
+
+	if b.proto3 && fieldDesc.Proto3Optional != nil && *fieldDesc.Proto3Optional {
+		idx := int32(len(b.messageDesc.Field)) - 1
+		fieldDesc.OneofIndex = &idx
+		name := "_" + *fieldDesc.Name
+		dcl := &pb.OneofDescriptorProto{Name: &name}
+		b.messageDesc.OneofDecl = append(b.messageDesc.OneofDecl, dcl)
+	}
+}
+
+var scalars = map[parser.Scalar]pb.FieldDescriptorProto_Type{
+	parser.Double:   pb.FieldDescriptorProto_TYPE_DOUBLE,
+	parser.Float:    pb.FieldDescriptorProto_TYPE_FLOAT,
+	parser.Int32:    pb.FieldDescriptorProto_TYPE_INT32,
+	parser.Int64:    pb.FieldDescriptorProto_TYPE_INT64,
+	parser.Uint32:   pb.FieldDescriptorProto_TYPE_UINT32,
+	parser.Uint64:   pb.FieldDescriptorProto_TYPE_UINT64,
+	parser.Sint32:   pb.FieldDescriptorProto_TYPE_SINT32,
+	parser.Sint64:   pb.FieldDescriptorProto_TYPE_SINT64,
+	parser.Fixed32:  pb.FieldDescriptorProto_TYPE_FIXED32,
+	parser.Fixed64:  pb.FieldDescriptorProto_TYPE_FIXED64,
+	parser.SFixed32: pb.FieldDescriptorProto_TYPE_SFIXED32,
+	parser.SFixed64: pb.FieldDescriptorProto_TYPE_SFIXED64,
+	parser.Bool:     pb.FieldDescriptorProto_TYPE_BOOL,
+	parser.String:   pb.FieldDescriptorProto_TYPE_STRING,
+	parser.Bytes:    pb.FieldDescriptorProto_TYPE_BYTES,
+}
+
+func (b *messageBuilder) createField(pf *parser.Field) *pb.FieldDescriptorProto {
+	if pf.Direct == nil || pf.Direct.Type.Map != nil {
+		panic("non-direct not implemented")
+	}
+
+	tag := int32(pf.Direct.Tag)
+	df := &pb.FieldDescriptorProto{
+		Name:           &pf.Direct.Name,
+		Number:         &tag,
+		JsonName:       jsonStr(pf.Direct.Name),
+		Label:          fieldLabel(pf, b.proto3),
+		Proto3Optional: proto3Optional(pf, b.proto3),
+	}
+
+	if pf.Direct.Type.Reference != nil {
+		t := pb.FieldDescriptorProto_TYPE_MESSAGE
+		df.Type = &t
+		df.TypeName = typeName(*pf.Direct.Type.Reference)
+		return df
+	}
+
+	fieldType := scalars[pf.Direct.Type.Scalar]
+	df.Type = &fieldType
+	return df
+}
+
+func fieldLabel(pf *parser.Field, proto3 bool) *pb.FieldDescriptorProto_Label {
+	var label pb.FieldDescriptorProto_Label
+	switch {
+	case pf.Required:
+		label = pb.FieldDescriptorProto_LABEL_REQUIRED
+	case pf.Repeated:
+		label = pb.FieldDescriptorProto_LABEL_REPEATED
+	case pf.Optional:
+		label = pb.FieldDescriptorProto_LABEL_OPTIONAL
+	case proto3:
+		// unlabelled proto3 field
+		label = pb.FieldDescriptorProto_LABEL_OPTIONAL
+	default:
+		panic("invalid field label for syntax")
```

Comment:
For all these panics on `default` I would do something like `panic(fmt.Sprintf("%s: invalid field label for syntax", pf.Pos))` otherwise it will be very difficult to debug when/if it occurs.

Reply:
TBH I actually usually just do `panic(pf.Pos.String())` because the stack trace usually gives you enough context to find the issue and I'm lazy.

Reply:
done.